src/YetAnotherMaximizedWETH.sol

```diff
@@ -390,4 +401,66 @@ contract YAM_WETH {
             }
         }
     }
+
+    /// @notice Sends a flash loan of `amount` WETH to the caller from the `from`
+    ///         address' balance.
+    /// @dev    Reverts if the caller does not repay their loan by the end of the
+    ///         subcontext's execution.
+    ///         Will be updated to use transient storage post EIP-1153.
+    /// @param  from   - The receiver of the flash loan. Must be a contract with a
+    ///                 `flashWeth(uint256)` function.
+    /// @param  amount - The amount of WETH in WEI to be loaned to `msg.sender`.
+    function flashLoan(address from, uint256 amount) external {
```

Comment:
Borrowing from existing balances is an interesting approach as it ensures the "can always redeem 1 WETH for 1 ETH" invariant is never broken but I think it breaks another more important invariant: "the token balance of an account cannot decrease without their implicit / explicit permission or action".

For example imagine a simple Uniswap V2 pool holding 1 WETH and 1000 DAI. Using `flashLoan` you borrow 0.5 WETH and call [`sync`](https://github.com/Uniswap/v2-core/blob/ee547b17853e71ed4e0101ccfd52e70d5acded58/contracts/UniswapV2Pair.sol#L198) on the pool pushing the price of 1 WETH to 2000 DAI (1000 DAI / 0.5 WETH). You then "return" the 0.5 WETH by swapping via the pool, receiving some DAI in return. 

Borrowing from other balances could still work if:
1. It's opt-in
2. Flashloan is repaid not by checking the lender's balance but by checking the borrower's balance and transferring

Point 2. prevents other forms of exploit where you e.g. buy an NFT from a seller, borrowing the WETH first and your purchase restores their balance, effectively stealing the asset.